### PR TITLE
Refactor KTrajectory type and FourierOP

### PR DESCRIPTION
**Pull request by fzimmermann89**
A suggestion for a refactor of some code.

- Trajectory Type is now a Flag with values singlevalue and ongrid.  'notongrid' is not 'ongrid'
- KTrajectory is now frozen, This causes issues in the wasabi example. We might reconsider making it unfrozen.
- Trajectory type is now cached (and inalidated if the tensor version cahnges because of some inplace operation)

- FourierOP uses reshape/permute etc in the nufft. The advantage of rearange gets lost of we have compicated code building the expression string.
- Steps toward allowing multiple other dimensions in the FourierOP
- I do not fully understand to NotImplemented Exception. IMHO this check could be wrong,

- Flake8 and Black disagree in E203. I disabled E203 in flake8

### CODE REVIEW

**Comment by fzimmermann89** [comment]
Precommit fails in mypy in the wasabi example because KTrajctory is now frozen and the example directly changes the Tensor.

Test fails in test_fourier_op.py::test_fourier_not_supported_traj, as I commented out this check for now. This needs a fix.

File: .gitignore [hunk 1]
```diff
@@ -161,3 +161,7 @@ cython_debug/
 
 # Documentation
 **/_autosummary
+
+# anaconda related directories
```
Comment (ckolbPTB) [inline .gitignore:165]: Why is this needed?

File: src/mrpro/data/_KTrajectory.py [hunk 2]
```diff
@@ -36,9 +36,18 @@ class KTrajectory:
...
+
+    ky: torch.Tensor
+    """(other,k2,k1,k0), phase encoding direction k1 if Cartesian."""
+
+    kx: torch.Tensor
+    """(other,k2,k1,k0), frequency encoding direction k0 if Cartesian."""
+
+    _type_along_kzyx: tuple[TrajType, TrajType, TrajType]
+    _type_along_k210: tuple[TrajType, TrajType, TrajType]
+    _expected_tensor_versions: tuple[int, int, int]
```
Comment (ckolbPTB) [inline src/mrpro/data/_KTrajectory.py]: I would remove `_expected_tensor_versions: tuple[int, int, int]` and always run `_traj_types()` whenever traj_type is queried. not efficient but should still be ok. Otherwise errors about changed tensors would occur in random places.

File: src/mrpro/data/_KTrajectory.py [hunk 3]
```diff
@@ -61,106 +70,60 @@ def broadcasted_shape(self) -> tuple[int, ...]:
...
-    def _type_of_traj(self) -> None:
-        """Calculate type of trajectory.
+    def _traj_types(self, tolerance) -> tuple[tuple[TrajType, TrajType, TrajType], tuple[TrajType, TrajType, TrajType]]:
+        """Calculate the trajectory type along kzkykx and k2k1k0.
 
-        This function checks if the entries of the trajectory along certain dimensions
+        Checks if the entries of the trajectory along certain dimensions
             - are of shape 1 -> TrajType.SINGLEVALUE
             - lie on a Cartesian grid -> TrajType.ONGRID
             - none of the above -> TrajType.NOTONGRID
```
Comment (ckolbPTB) [inline src/mrpro/data/_KTrajectory.py]: TrajType.NOTONGRID does not exist anymore

File: src/mrpro/data/_KTrajectory.py [hunk 4]
```diff
@@ -61,106 +70,60 @@ def broadcasted_shape(self) -> tuple[int, ...]:
...
+        return (type_zyx[0], type_zyx[1], type_zyx[2]), (type_210[0], type_210[1], type_210[2])
 
-        Parameters
-        ----------
-        tolerance
-            tolerance used to determine if trajectory points lie on a gird
+    @property
+    def type_along_kzyx(self) -> tuple[TrajType, TrajType, TrajType]:
+        """Type of trajectory along kz-ky-kx."""
+        if self._expected_tensor_versions != (self.kz._version, self.ky._version, self.kx._version):
```
Comment (ckolbPTB) [inline src/mrpro/data/_KTrajectory.py]: remove - see comment above

File: src/mrpro/data/_KTrajectory.py [hunk 5]
```diff
@@ -61,106 +70,60 @@ def broadcasted_shape(self) -> tuple[int, ...]:
...
 
-        Returns
-        -------
-            type of trajectory along k2-k1-k0
-        """
-        return self._type_k2_k1_k0
+    @property
+    def type_along_k210(self) -> tuple[TrajType, TrajType, TrajType]:
+        """Type of trajectory along k2-k1-k0."""
+        if self._expected_tensor_versions != (self.kz._version, self.ky._version, self.kx._version):
```
Comment (ckolbPTB) [inline src/mrpro/data/_KTrajectory.py]: remove - see comment above

File: src/mrpro/data/_KTrajectory.py [hunk 6]
```diff
@@ -212,9 +173,19 @@ def __init__(
         if len(shape) != 4:
             raise ValueError('The k-space trajectory tensors should each have 4 dimensions.')
 
+        type_along_kzyx, type_along_k210 = self._traj_types(grid_detection_tolerance)
```
Comment (ckolbPTB) [inline src/mrpro/data/_KTrajectory.py]: not needed here

File: .gitignore (hunk 1 above)
Comment (fzimmermann89) [reply .gitignore:165]: We can also remove it. I put it there because I use a vscode extension that creates these folders in the project folder, thus they always show up in git status etc.
I am pretty sure we will never want to add a .conda environment to git, so adding it in the .gitignore along all the other ignored stuff will not hurt.

File: src/mrpro/operators/_FourierOp.py [hunk 7]
```diff
@@ -58,38 +56,46 @@ def __init__(
...
-            # E.g. kx is of shape (1,1,30,1,1): kx sampling along k2.
-            if self._fft_dims != [dim for dim in (-3, -2, -1) if traj.traj_type_along_k210[dim] == TrajType.ONGRID]:
+        if self._nufft_dims:
+            fft_dims_k210 = [
+                dim
+                for dim in (-3, -2, -1)
+                if (traj.type_along_k210[dim] & TrajType.ONGRID)
+                and not (traj.type_along_k210[dim] & TrajType.SINGLEVALUE)
+            ]
+            if self._fft_dims != fft_dims_k210:
```
Comment (ckolbPTB) [inline src/mrpro/operators/_FourierOp.py:90]: This checked here because as far as I can see, there is only a problem if the FFT dimensions are not aligned and a combination of FFT and NUFFT is carried out. Maybe we can rephrase the error message to make this clearer.

File: .gitignore (hunk 1 above)
Comment (ckolbPTB) [reply .gitignore:165]: Our .gitignore is really big anyway so probably doesn't matter. I prefer to use something like .*/ to e.g. ignore all folders starting with a period and then adding specific folders which are needed. Makes it easier to read and maintain. Maybe for the future...

**Comment by ckolbPTB** [comment]
@fzimmermann89 let me know if I have missed anything otherwise feel free to merge
